File: apps/cli/src/commands/scaffold.ts
/**
 * `frick scaffold <kind> <name>` — append a stub to an already-initialized
 * app. Objects and streams use PascalCase names; projections use kebab-case.
 *
 * Objects and streams append a typed literal between the `// frick:objects`
 * or `// frick:streams` markers in `src/schema.ts`. Projections create a
 * new file under `src/projections/<name>.ts` and wire it into `src/server.ts`
 * via the `// frick:projections:imports` and `// frick:projections:register`
 * markers.
 *
 * All three subcommands are idempotent: a second call with the same name is
 * refused with exit 3 rather than silently duplicating.
 */
import { existsSync } from "node:fs";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { dirname, isAbsolute, join, resolve } from "node:path";
import type { ParsedArgs } from "../argv.js";
import { CliRefusedError, CliUsageError } from "../errors.js";
import { emit, type OutputOptions } from "../output.js";

const OBJECT_MARKER = "// frick:objects";
const STREAM_MARKER = "// frick:streams";
const PROJECTION_IMPORTS_MARKER = "// frick:projections:imports";
const PROJECTION_REGISTER_MARKER = "// frick:projections:register";

function resolveDirectory(parsed: ParsedArgs): string {
  const flag = parsed.flags.directory ?? parsed.flags.cwd;
  if (typeof flag === "string" && flag.length > 0) {
    return isAbsolute(flag) ? flag : resolve(process.cwd(), flag);
  }
  return process.cwd();
}

function requirePositional(parsed: ParsedArgs, index: number, label: string): string {
  const value = parsed.positionals[index];
  if (!value) {
    throw new CliUsageError(`frick scaffold requires ${label}`);
  }
  return value;
}

function assertPascalCase(name: string, label: string): void {
  if (!/^[A-Z][A-Za-z0-9]*$/.test(name)) {
    throw new CliUsageError(`${label} must be PascalCase, got ${JSON.stringify(name)}`);
  }
}

function assertKebabCase(name: string, label: string): void {
  if (!/^[a-z][a-z0-9-]*$/.test(name)) {
    throw new CliUsageError(`${label} must be kebab-case, got ${JSON.stringify(name)}`);
  }
}

async function readSchemaFile(directory: string): Promise<{ path: string; body: string }> {
  const path = join(directory, "src", "schema.ts");
  if (!existsSync(path)) {
    throw new CliRefusedError(
      `No src/schema.ts found in ${directory}. Run 'frick init' first.`,
      { path },
    );
  }
  return { path, body: await readFile(path, "utf8") };
}

function insertAfterMarker(body: string, marker: string, insertion: string): string {
  const idx = body.indexOf(marker);
  if (idx < 0) {
    throw new CliRefusedError(`schema.ts is missing marker ${marker}`);
  }
  // Find end of the marker line, then insert insertion immediately after.
  const lineEnd = body.indexOf("\n", idx);
  if (lineEnd < 0) {
    throw new CliRefusedError(`schema.ts marker ${marker} not at start of a line`);
  }
  return `${body.slice(0, lineEnd + 1)}${insertion}${body.slice(lineEnd + 1)}`;
}

function nextNumericId(body: string, sectionLabel: "objects" | "streams"): number {
  // Heuristic: count existing entries by counting `id: <n>,` near the top of
  // each appended block. We keep IDs monotonic per-section by counting how
  // many of our stub markers appear. This is a deliberate shortcut — the
  // developer is expected to renumber if they care; the framework validates
  // identity at runtime.
  const re = new RegExp(`// frick:${sectionLabel}:id (\\d+)`, "g");
  let maxId = 0;
  for (const match of body.matchAll(re)) {
    const n = Number(match[1]);
    if (n > maxId) maxId = n;
  }
  return maxId + 1;
}

function objectStub(name: string, id: number): string {
  return `  // frick:objects:id ${id} ${name}
  {
    id: ${id},
    name: ${JSON.stringify(name)},
    fields: [
      { id: 1, name: "displayName", kind: "string", required: true },
    ],
    indexes: [{ id: 1, name: "all", fields: ["displayName"] }],
  },
`;
}

function streamStub(name: string, id: number): string {
  return `  // frick:streams:id ${id} ${name}
  {
    id: ${id},
    name: ${JSON.stringify(name)},
    keyFields: [{ id: 1, name: "key", kind: "string", required: true }],
    events: [],
  },
`;
}

function appendToArrayLiteral(body: string, sectionName: "objects" | "streams", insertion: string): string {
  // Find `<section>: [` and the matching `]` to splice insertion at its tail.
  const header = `${sectionName}: [`;
  const headerIdx = body.indexOf(header);
  if (headerIdx < 0) {
    throw new CliRefusedError(`schema.ts is missing '${header}' literal`);
  }
  // Find the matching closing bracket. The scaffolded schema uses simple
  // shallow arrays so naive scanning is sufficient; if a future template
  // nests brackets inside the section, this needs upgrading.
  let depth = 0;
  let i = headerIdx + header.length - 1; // position at the '['
  for (; i < body.length; i++) {
    const ch = body[i];
    if (ch === "[") depth += 1;
    else if (ch === "]") {
      depth -= 1;
      if (depth === 0) break;
    }
  }
  if (depth !== 0) {
    throw new CliRefusedError(`schema.ts '${sectionName}' literal is unbalanced`);
  }
  // Insert before the closing bracket. Add a leading newline if the array is
  // currently empty (i.e. literal "[]").
  const before = body.slice(0, i);
  const after = body.slice(i);
  const needsNewlineBefore = before.endsWith("[");
  const prefix = needsNewlineBefore ? "\n" : "";
  return `${before}${prefix}${insertion}${after}`;
}

function sectionContainsName(body: string, sectionLabel: "objects" | "streams", name: string): boolean {
  const re = new RegExp(`// frick:${sectionLabel}:id \\d+ ${name}\\b`);
  return re.test(body);
}

async function scaffoldObject(parsed: ParsedArgs, out: OutputOptions): Promise<number> {
  const name = requirePositional(parsed, 1, "<Name> for the object");
  assertPascalCase(name, "object name");
  const directory = resolveDirectory(parsed);
  const { path, body } = await readSchemaFile(directory);
  if (sectionContainsName(body, "objects", name)) {
    throw new CliRefusedError(`Object ${name} already exists in schema.ts`, { path });
  }
  const id = nextNumericId(body, "objects");
  const next = appendToArrayLiteral(body, "objects", objectStub(name, id));
  await writeFile(path, next, "utf8");
  emit({ ok: true, kind: "object", name, id, path }, out);
  return 0;
}

async function scaffoldStream(parsed: ParsedArgs, out: OutputOptions): Promise<number> {
  const name = requirePositional(parsed, 1, "<Name> for the stream");
  assertPascalCase(name, "stream name");
  const directory = resolveDirectory(parsed);
  const { path, body } = await readSchemaFile(directory);
  if (sectionContainsName(body, "streams", name)) {
    throw new CliRefusedError(`Stream ${name} already exists in schema.ts`, { path });
  }
  const id = nextNumericId(body, "streams");
  const next = appendToArrayLiteral(body, "streams", streamStub(name, id));
  await writeFile(path, next, "utf8");
  emit({ ok: true, kind: "stream", name, id, path }, out);
  return 0;
}

function projectionFile(name: string): string {
  // The handler is intentionally minimal — apps fill in apply()/rebuild() once
  // they know which stream they're consuming. We deliberately do not import
  // from @fricken/server here because that would pin the scaffolded app to a
  // single server-internal API surface; the developer wires the types when
  // they implement the handler.
  return `/**
 * Projection scaffold for "${name}". Wire this into your server via the
 * registry returned from createFrickServer.
 */
export function create${toPascalCase(name)}Projection() {
  return {
    name: ${JSON.stringify(name)},
    sources: [] as const,
    handler: {
      apply(_event: unknown, _ctx: unknown) {
        return { changes: [] as unknown[] };
      },
    },
  };
}
`;
}

function toPascalCase(kebab: string): string {
  return kebab
    .split("-")
    .filter((part) => part.length > 0)
    .map((part) => part[0]!.toUpperCase() + part.slice(1))
    .join("");
}

async function readServerFile(directory: string): Promise<{ path: string; body: string }> {
  const path = join(directory, "src", "server.ts");
  if (!existsSync(path)) {
    throw new CliRefusedError(
      `No src/server.ts found in ${directory}. Run 'frick init' first.`,
      { path },
    );
  }
  return { path, body: await readFile(path, "utf8") };
}

async function scaffoldProjection(parsed: ParsedArgs, out: OutputOptions): Promise<number> {
  const name = requirePositional(parsed, 1, "<name> for the projection (kebab-case)");
  assertKebabCase(name, "projection name");
  const directory = resolveDirectory(parsed);

  const projectionPath = join(directory, "src", "projections", `${name}.ts`);
  if (existsSync(projectionPath)) {
    throw new CliRefusedError(`Projection file already exists: ${projectionPath}`, {
      path: projectionPath,
    });
  }

  const { path: serverPath, body: serverBody } = await readServerFile(directory);
  const factoryName = `create${toPascalCase(name)}Projection`;
  if (serverBody.includes(factoryName)) {
    throw new CliRefusedError(`Projection ${name} appears already registered in server.ts`, {
      path: serverPath,
    });
  }

  await mkdir(dirname(projectionPath), { recursive: true });
  await writeFile(projectionPath, projectionFile(name), "utf8");

  const importLine = `import { ${factoryName} } from "./projections/${name}.js";\n`;
  // The scaffolded server treats projections as opt-in: we emit a reference
  // so the import is not dead code, and a TODO so the developer wires the
  // factory into their app's projection registry when they're ready.
  const registerLine = `// TODO: register ${factoryName}() with your projection registry\nvoid ${factoryName};\n`;

  let nextServer = insertAfterMarker(serverBody, PROJECTION_IMPORTS_MARKER, importLine);
  nextServer = insertAfterMarker(nextServer, PROJECTION_REGISTER_MARKER, registerLine);
  await writeFile(serverPath, nextServer, "utf8");

  emit(
    {
      ok: true,
      kind: "projection",
      name,
      projectionPath,
      serverPath,
    },
    out,
  );
  return 0;
}

export async function scaffoldCommand(parsed: ParsedArgs, out: OutputOptions): Promise<number> {
  const sub = parsed.positionals[0];
  if (sub === "object") return scaffoldObject(parsed, out);
  if (sub === "stream") return scaffoldStream(parsed, out);
  if (sub === "projection") return scaffoldProjection(parsed, out);
  throw new CliUsageError(`Unknown scaffold subcommand: ${sub ?? "<missing>"}`, {
    expected: ["object", "stream", "projection"],
  });
}

File: crates/frick-cli/src/commands/scaffold.rs
//! `frick scaffold object|stream|projection <Name>`
//! (ported from `apps/cli/src/commands/scaffold.ts`).
//!
//! Objects/streams: PascalCase names, spliced into the `objects: [` / `streams:
//! [` array literal in `src/schema.ts`, IDs allocated from `// frick:<sec>:id N`
//! markers. Projections: kebab-case names, a new `src/projections/<name>.ts`
//! plus import/register lines in `src/server.ts`. Duplicate ⇒ refusal (exit 3).

use std::fs;
use std::path::{Path, PathBuf};

use serde_json::json;

use crate::argv::ParsedArgs;
use crate::errors::{CliError, EXIT_OK};
use crate::output::Output;

const PROJECTION_IMPORTS_MARKER: &str = "// frick:projections:imports";
const PROJECTION_REGISTER_MARKER: &str = "// frick:projections:register";

/// `scaffoldCommand`.
pub fn scaffold_command(parsed: &ParsedArgs, out: &mut Output) -> Result<i32, CliError> {
    match parsed.positional(0) {
        Some("object") => scaffold_object(parsed, out),
        Some("stream") => scaffold_stream(parsed, out),
        Some("projection") => scaffold_projection(parsed, out),
        other => Err(CliError::usage_with(
            format!(
                "Unknown scaffold subcommand: {}",
                other.unwrap_or("<missing>")
            ),
            json!({ "expected": ["object", "stream", "projection"] }),
        )),
    }
}

fn resolve_directory(parsed: &ParsedArgs) -> PathBuf {
    let flag = parsed
        .flag_str("directory")
        .or_else(|| parsed.flag_str("cwd"));
    match flag {
        Some(value) if !value.is_empty() => {
            let path = Path::new(value);
            if path.is_absolute() {
                path.to_path_buf()
            } else {
                std::env::current_dir()
                    .unwrap_or_else(|_| PathBuf::from("."))
                    .join(path)
            }
        }
        _ => std::env::current_dir().unwrap_or_else(|_| PathBuf::from(".")),
    }
}

fn require_positional<'a>(
    parsed: &'a ParsedArgs,
    index: usize,
    label: &str,
) -> Result<&'a str, CliError> {
    parsed
        .positional(index)
        .filter(|v| !v.is_empty())
        .ok_or_else(|| CliError::usage(format!("frick scaffold requires {label}")))
}

fn assert_pascal_case(name: &str, label: &str) -> Result<(), CliError> {
    let mut chars = name.chars();
    let valid = matches!(chars.next(), Some(c) if c.is_ascii_uppercase())
        && chars.all(|c| c.is_ascii_alphanumeric());
    if valid {
        Ok(())
    } else {
        Err(CliError::usage(format!(
            "{label} must be PascalCase, got {}",
            serde_json::to_string(name).unwrap_or_else(|_| format!("\"{name}\""))
        )))
    }
}

fn assert_kebab_case(name: &str, label: &str) -> Result<(), CliError> {
    let mut chars = name.chars();
    let valid = matches!(chars.next(), Some(c) if c.is_ascii_lowercase())
        && chars.all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '-');
    if valid {
        Ok(())
    } else {
        Err(CliError::usage(format!(
            "{label} must be kebab-case, got {}",
            serde_json::to_string(name).unwrap_or_else(|_| format!("\"{name}\""))
        )))
    }
}

fn read_schema_file(directory: &Path) -> Result<(PathBuf, String), CliError> {
    let path = directory.join("src").join("schema.ts");
    if !path.exists() {
        return Err(CliError::refused_with(
            format!(
                "No src/schema.ts found in {}. Run 'frick init' first.",
                directory.display()
            ),
            json!({ "path": path.to_string_lossy() }),
        ));
    }
    let body = fs::read_to_string(&path)
        .map_err(|err| CliError::failure("cli.io", format!("read failed: {err}")))?;
    Ok((path, body))
}

fn insert_after_marker(body: &str, marker: &str, insertion: &str) -> Result<String, CliError> {
    let Some(idx) = body.find(marker) else {
        return Err(CliError::refused(format!(
            "schema.ts is missing marker {marker}"
        )));
    };
    let Some(line_end) = body[idx..].find('\n').map(|rel| idx + rel) else {
        return Err(CliError::refused(format!(
            "schema.ts marker {marker} not at start of a line"
        )));
    };
    Ok(format!(
        "{}{insertion}{}",
        &body[..=line_end],
        &body[line_end + 1..]
    ))
}

fn next_numeric_id(body: &str, section_label: &str) -> i64 {
    let needle = format!("// frick:{section_label}:id ");
    let mut max_id = 0_i64;
    let mut search = body;
    while let Some(pos) = search.find(&needle) {
        let after = &search[pos + needle.len()..];
        let digits: String = after.chars().take_while(char::is_ascii_digit).collect();
        if let Ok(n) = digits.parse::<i64>()
            && n > max_id
        {
            max_id = n;
        }
        search = &after[digits.len()..];
    }
    max_id + 1
}

fn object_stub(name: &str, id: i64) -> String {
    let name_lit = serde_json::to_string(name).unwrap_or_else(|_| format!("\"{name}\""));
    format!(
        "  // frick:objects:id {id} {name}\n\
  {{\n\
    id: {id},\n\
    name: {name_lit},\n\
    fields: [\n\
      {{ id: 1, name: \"displayName\", kind: \"string\", required: true }},\n\
    ],\n\
    indexes: [{{ id: 1, name: \"all\", fields: [\"displayName\"] }}],\n\
  }},\n"
    )
}

fn stream_stub(name: &str, id: i64) -> String {
    let name_lit = serde_json::to_string(name).unwrap_or_else(|_| format!("\"{name}\""));
    format!(
        "  // frick:streams:id {id} {name}\n\
  {{\n\
    id: {id},\n\
    name: {name_lit},\n\
    keyFields: [{{ id: 1, name: \"key\", kind: \"string\", required: true }}],\n\
    events: [],\n\
  }},\n"
    )
}

fn append_to_array_literal(
    body: &str,
    section_name: &str,
    insertion: &str,
) -> Result<String, CliError> {
    let header = format!("{section_name}: [");
    let Some(header_idx) = body.find(&header) else {
        return Err(CliError::refused(format!(
            "schema.ts is missing '{header}' literal"
        )));
    };
    let bytes = body.as_bytes();
    let mut depth = 0_i32;
    let mut i = header_idx + header.len() - 1; // position at the '['
    let mut closing: Option<usize> = None;
    while i < bytes.len() {
        match bytes[i] {
            b'[' => depth += 1,
            b']' => {
                depth -= 1;
                if depth == 0 {
                    closing = Some(i);
                    break;
                }
            }
            _ => {}
        }
        i += 1;
    }
    let Some(close_idx) = closing else {
        return Err(CliError::refused(format!(
            "schema.ts '{section_name}' literal is unbalanced"
        )));
    };
    let before = &body[..close_idx];
    let after = &body[close_idx..];
    let prefix = if before.ends_with('[') { "\n" } else { "" };
    Ok(format!("{before}{prefix}{insertion}{after}"))
}

fn section_contains_name(body: &str, section_label: &str, name: &str) -> bool {
    // Mirror the TS regexp `// frick:<sec>:id \d+ <name>\b`.
    let prefix = format!("// frick:{section_label}:id ");
    let mut search = body;
    while let Some(pos) = search.find(&prefix) {
        let after = &search[pos + prefix.len()..];
        let digits: String = after.chars().take_while(char::is_ascii_digit).collect();
        let rest = &after[digits.len()..];
        if let Some(tail) = rest.strip_prefix(' ')
            && tail.starts_with(name)
        {
            // `\b` after the name: next char must not be alphanumeric/underscore.
            let next = tail[name.len()..].chars().next();
            if next.is_none_or(|c| !(c.is_ascii_alphanumeric() || c == '_')) {
                return true;
            }
        }
        search = &after[digits.len()..];
    }
    false
}

fn scaffold_object(parsed: &ParsedArgs, out: &mut Output) -> Result<i32, CliError> {
    let name = require_positional(parsed, 1, "<Name> for the object")?.to_string();
    assert_pascal_case(&name, "object name")?;
    let directory = resolve_directory(parsed);
    let (path, body) = read_schema_file(&directory)?;
    if section_contains_name(&body, "objects", &name) {
        return Err(CliError::refused_with(
            format!("Object {name} already exists in schema.ts"),
            json!({ "path": path.to_string_lossy() }),
        ));
    }
    let id = next_numeric_id(&body, "objects");
    let next = append_to_array_literal(&body, "objects", &object_stub(&name, id))?;
    fs::write(&path, next)
        .map_err(|err| CliError::failure("cli.io", format!("write failed: {err}")))?;
    out.emit(&json!({
        "ok": true,
        "kind": "object",
        "name": name,
        "id": id,
        "path": path.to_string_lossy(),
    }));
    Ok(EXIT_OK)
}

fn scaffold_stream(parsed: &ParsedArgs, out: &mut Output) -> Result<i32, CliError> {
    let name = require_positional(parsed, 1, "<Name> for the stream")?.to_string();
    assert_pascal_case(&name, "stream name")?;
    let directory = resolve_directory(parsed);
    let (path, body) = read_schema_file(&directory)?;
    if section_contains_name(&body, "streams", &name) {
        return Err(CliError::refused_with(
            format!("Stream {name} already exists in schema.ts"),
            json!({ "path": path.to_string_lossy() }),
        ));
    }
    let id = next_numeric_id(&body, "streams");
    let next = append_to_array_literal(&body, "streams", &stream_stub(&name, id))?;
    fs::write(&path, next)
        .map_err(|err| CliError::failure("cli.io", format!("write failed: {err}")))?;
    out.emit(&json!({
        "ok": true,
        "kind": "stream",
        "name": name,
        "id": id,
        "path": path.to_string_lossy(),
    }));
    Ok(EXIT_OK)
}

fn to_pascal_case(kebab: &str) -> String {
    kebab
        .split('-')
        .filter(|part| !part.is_empty())
        .map(|part| {
            let mut chars = part.chars();
            chars.next().map_or_else(String::new, |first| {
                first.to_ascii_uppercase().to_string() + chars.as_str()
            })
        })
        .collect()
}

fn projection_file(name: &str) -> String {
    let name_lit = serde_json::to_string(name).unwrap_or_else(|_| format!("\"{name}\""));
    let pascal = to_pascal_case(name);
    format!(
        "/**\n\
 * Projection scaffold for \"{name}\". Register it in src/server.ts — it is\n\
 * added to the app's `projections` array via the marker comments.\n\
 */\n\
export function create{pascal}Projection() {{\n\
  return {{\n\
    name: {name_lit},\n\
    sources: [] as const,\n\
    handler: {{\n\
      apply(_event: unknown, _ctx: unknown) {{\n\
        return {{ changes: [] as unknown[] }};\n\
      }},\n\
    }},\n\
  }};\n\
}}\n"
    )
}

fn read_server_file(directory: &Path) -> Result<(PathBuf, String), CliError> {
    let path = directory.join("src").join("server.ts");
    if !path.exists() {
        return Err(CliError::refused_with(
            format!(
                "No src/server.ts found in {}. Run 'frick init' first.",
                directory.display()
            ),
            json!({ "path": path.to_string_lossy() }),
        ));
    }
    let body = fs::read_to_string(&path)
        .map_err(|err| CliError::failure("cli.io", format!("read failed: {err}")))?;
    Ok((path, body))
}

fn scaffold_projection(parsed: &ParsedArgs, out: &mut Output) -> Result<i32, CliError> {
    let name = require_positional(parsed, 1, "<name> for the projection (kebab-case)")?.to_string();
    assert_kebab_case(&name, "projection name")?;
    let directory = resolve_directory(parsed);

    let projection_path = directory
        .join("src")
        .join("projections")
        .join(format!("{name}.ts"));
    if projection_path.exists() {
        return Err(CliError::refused_with(
            format!(
                "Projection file already exists: {}",
                projection_path.display()
            ),
            json!({ "path": projection_path.to_string_lossy() }),
        ));
    }

    let (server_path, server_body) = read_server_file(&directory)?;
    let factory_name = format!("create{}Projection", to_pascal_case(&name));
    if server_body.contains(&factory_name) {
        return Err(CliError::refused_with(
            format!("Projection {name} appears already registered in server.ts"),
            json!({ "path": server_path.to_string_lossy() }),
        ));
    }

    if let Some(parent) = projection_path.parent() {
        fs::create_dir_all(parent)
            .map_err(|err| CliError::failure("cli.io", format!("mkdir failed: {err}")))?;
    }
    fs::write(&projection_path, projection_file(&name))
        .map_err(|err| CliError::failure("cli.io", format!("write failed: {err}")))?;

    let import_line = format!("import {{ {factory_name} }} from \"./projections/{name}.js\";\n");
    let register_line = format!(
        "// TODO: register {factory_name}() with your projection registry\nvoid {factory_name};\n"
    );

    let mut next_server =
        insert_after_marker(&server_body, PROJECTION_IMPORTS_MARKER, &import_line)?;
    next_server = insert_after_marker(&next_server, PROJECTION_REGISTER_MARKER, &register_line)?;
    fs::write(&server_path, next_server)
        .map_err(|err| CliError::failure("cli.io", format!("write failed: {err}")))?;

    out.emit(&json!({
        "ok": true,
        "kind": "projection",
        "name": name,
        "projectionPath": projection_path.to_string_lossy(),
        "serverPath": server_path.to_string_lossy(),
    }));
    Ok(EXIT_OK)
}
